shared/reducers/chat.js

```diff
@@ -25,6 +25,34 @@ function dedupeMessages (seenMessages: Set<MessageID>, knownMessages: List<Serve
   }
 }
 
+type ConversationsStates = Map<Constants.ConversationIDKey, ConversationState>
+type ConversationUpdateFn = (c: Constants.ConversationState) => Constants.ConversationState
+function updateConversation (conversationStates: ConversationsStates, conversationIDKey: Constants.ConversationIDKey, conversationUpdateFn: ConversationUpdateFn): ConversationsStates {
+  return conversationStates.update(
+    conversationIDKey,
+    initialConversation,
+    conversationUpdateFn,
+  )
+}
+
+type MessageUpdateFn = (message: $Subtype<Message>) => Message
+type MessageFindPredFn = (message: Message) => boolean
+function updateConversationMessage (conversationStates: ConversationsStates, conversationIDKey: Constants.ConversationIDKey, pred: MessageFindPredFn, messageUpdateFn: MessageUpdateFn): ConversationsStates {
+  return updateConversation(
+    conversationStates,
+    conversationIDKey,
+    conversation => {
+      const index = conversation.get('messages').findIndex(pred)
+      if (index < 0) {
+        console.warn("Couldn't find an outbox entry to modify")
```

Comment:
I wonder if we should throw an error here...

Console warns are kind of only appropriate for warnings?